experiments/process_sandbox/README.md

```diff
@@ -330,3 +329,184 @@ Code layout
+
+* [test-sandbox-basic](tests/sandbox-basic.cc): Creates 3 sandboxes wrapping around a small library that exposes a function returning the sum of two `int`. Each sandbox is called twice.
+
+* [test-sandbox-callback-basic](tests/sandbox-callback-basic.cc): Creates a simple sandbox and registers a callback function that returns an `int`. The sandbox library simply invokes the callback and asserts it gets the correct value (`42`). 
+
+* [test-sandbox-callback-recursive](tests/sandbox-callback-recursive.cc): Creates a single sandbox that recursively invokes 6 nested levels of callbacks, i.e., a callback triggers a call into the sandbox that itself invokes a callback etc. 
+
+* [test-sandbox-crash](tests/sandbox-crash.cc): Creates 3 sandboxes meant to crash (calls `abort`). The crash triggers a `runtime_error` exception caught in the parent process. 
+
+* [test-sandbox-fake-open](tests/sandbox-fake-open.cc): Creates a shared memory file and a sandbox. The sandbox opens the file in two different ways: 1) with a raw system call, and 2) with the `open` function that performs a callback to the host. 
+
+* [test-sandbox-modify-pagemap](tests/sandbox-modify-pagemap.cc): Creates a sandbox attempting to modify the access rights of the pagemap using `mprotect`. The `mprotect` call should fail with a `permission denied`.
+
+* [test-sandbox-zlib](tests/sandbox-zlib.cc): Uses `zlib` to `deflate`, i.e., compress, the program's file. This test compares the output produced by a sandboxed and an unsandboxed version of zlib. 
+
+* [test-sandbox-network](tests/test-sandbox-network.cc): Example of how network policies can be used to restrict the sandbox's ability to perform network operations. 
+This test creates a server sandbox and gradually enables network operations (1) `getaddrinfo`, (2) `bind`ing to a socket, (3) the ability to receive messages on the socket from the parent, and (4) the ability for asandboxed client to `connect` to the server. 
+
+Walkthrough Example
+-------------------
+
+In this section we provide a slightly more detailed walkthrough for the `test-sandbox-basic` test.
+
+### Files
+
+* [tests/sandbox-basic.cc](tests/sandbox-basic.cc): Defines the main entry point of the test.
+* [tests/sandboxlib-basic.cc](tests/sandboxlib-basic.cc): Defines the basic library that will be sandboxed.
+
+### Running the test
+
+```
+./build_ninja/tests/test-sandbox-basic
+
+# expected Output
+
+Adding 1 to 2 in sandbox
+Adding 2147483647 to -1 in sandbox
+Adding 1 to 2 in sandbox
+Adding 2147483647 to -1 in sandbox
+Adding 1 to 2 in sandbox
+Adding 2147483647 to -1 in sandbox
+```
+
+### High Level Description
+
+This tests creates 3 sandboxes wrapping around the simple adder defined by the sandboxed library.
+Each sandbox runs in a separate process and is called twice.
+
+
+### Runtime Walkthrough
+
+Below is the process tree generated by the test:
+
+```
+test-sandbox-basic─┬─3*[library_runner]
+                   └─{test-sandbox-basic}
```

Comment:
What is this line?  It looks like `test-sandbox-basic` is a child of itself?

Reply:
This is the MemoryServiceProvider:
https://github.com/microsoft/verona/blob/master/experiments/process_sandbox/src/libsandbox.cc#L280